oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationEndpointFilter.java

```diff
@@ -717,12 +829,14 @@ private static boolean isConsentCancelled(HttpServletRequest request) {
 		}
 
 		private static String generateConsentPage(HttpServletRequest request,
-				RegisteredClient registeredClient, OAuth2Authorization authorization) {
-
+				RegisteredClient registeredClient, OAuth2Authorization authorization, OAuth2AuthorizationConsent previousConsent) {
 			OAuth2AuthorizationRequest authorizationRequest = authorization.getAttribute(
 					OAuth2AuthorizationRequest.class.getName());
 			Set<String> scopes = new HashSet<>(authorizationRequest.getScopes());
 			scopes.remove(OidcScopes.OPENID);		// openid scope does not require consent
+			if (previousConsent != null) {
+				scopes.removeAll(previousConsent.getScopes());
```

Comment:
We could, but then they should be disabled, because otherwise the users may think they are _revoking_ consent, or _not granting consent_ for this particular token request, despite them having not granted consent before.

The more I think about it, the more I think it should be visible but grayed-out.

---

Example flow if we just keep it as is.

### Token request #1:

User sees `message.read` and `message.write`, and only ticks the second:

- [ ] `message.read`
- [x] `message.write`

Then, consent for `message.write` is persisted. There is no way to revoke consent yet.

### Token request #2

User sees `message.read` and `message.write` again, and this times ticks only the first:

- [x] `message.read`
- [ ] `message.write`

The user would expect to _only_ consent to `message.read`. But they have already consented to `message.write`, so despite them unticking `message.write`, the client app will obtain a token with the `message.write` scope as requested.

Reply:
Done, please review and/or change the "previously approved scopes" message:

> You have already granted the following permissions to the above app: